### PR TITLE
improve conflict detection for binary files

### PR DESCRIPTION
**Pull request by shiftkey**
## Overview

As binary files don't contain conflicted markers, we need a different technique to detect when these are "resolved" in Git - the command line seems to prefer staging the change, so maybe we need to do that?

~~🌵🌵🌵 **Depends on #6074 and #6072, go review those first**🌵🌵🌵~~

**Relates to #6062**

## Description

 - [x] add failing test that incorrectly marks binary file as resolved
 - [x] upgrade `getStatus` so that it reports the binary file is conflicted
 - [x] rebase branch ontop of `master` once #6072 and #6074 are merged, so the diff just focuses on the necessary Git changes
 - [x] identify other test cases that we need to handle (added/removed/updated combinations) and add tests to verify those are detected

## Release notes

Notes: no-notes

### CODE REVIEW

**Comment by shiftkey** [comment]
For reference, the command line flow is a bit tricky:

```
$ git status 
On branch make-a-change
You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Unmerged paths:
  (use "git add <file>..." to mark resolution)

	both modified:   my-cool-image.png

no changes added to commit (use "git add" and/or "git commit -a")
```

I then `git add my-cool-image.png` to indicate I want `ours` as part of the merge, but then it disappears from the `status` output:

```
$ git status 
On branch make-a-change
All conflicts fixed but you are still merging.
  (use "git commit" to conclude merge)
```

Maybe that's fine for the purposes of this workflow?

**Comment by outofambit** [comment]
I think we could do a standalone pr for a `lib` module to detect if a file is a binary or not. what do you think @shiftkey? (I'd like to help get this figured out and I want to make sure we can split up work sensibly)

**Comment by shiftkey** [comment]
> I think we could do a standalone pr for a `lib` module to detect if a file is a binary or not. 

This is my primary focus right now, and I'll update this branch once I have things to show off. Feel free to dig into things from the UI end as I haven't started on that area yet.

**Comment by shiftkey** [comment]
@outofambit I've opened https://github.com/desktop/desktop/pull/6072 to get the right abstraction in place so that you can work on the UI for this.

**Comment by outofambit** [comment]
@shiftkey thank you!

**Comment by shiftkey** [comment]
Alright, I think this is ready for a first round of review. I'd like to get feedback on these parts:

 - do the changes make sense? I've had to incorporate a second `git` call to test for binary detection, and so there's now more details floating around in `getStatus` to understand
 - there's some new types in `models/conflicts.ts` that I've had to define as part of enhancing conflict detection - do they fit with your mental model of how Git conflicts work?
 - have I missed anything?

There's two tests for different binary conflicts currently, and I plan to add in some more tomorrow but the setup is tedious and I wanted to to focus on the code changes first.

File: app/src/lib/git/status.ts [hunk 1]
```diff
@@ -138,16 +158,29 @@ export async function getStatus(
...
-  )
-    ? await getFilesWithConflictMarkers(repository.path)
-    : new Map<string, number>()
+  const filesAndKeys = entries.map(es => ({
+    path: es.path,
+    status: mapStatus(es.statusCode),
+  }))
+
+  // implicit cast here because filter doesn't infer the union type correctly :(
+  const conflictedFilesInIndex = filesAndKeys.filter(
```
[the remark this answers is not in the document]
Comment (shiftkey) [reply app/src/lib/git/status.ts]: Yessss, I thought we had something similiar lying around. Good spot!

**Comment by shiftkey** [comment]
@outofambit yeah, i'm keen to see if we can keep moving towards an application-specific conflicts interface that makes sense. Happy to keep nudging things along while we're wrapping up 1.5.0.

**Comment by shiftkey** [comment]
@outofambit just had to resolve a conflict from #6081 that prevented this from being merge

**Pull request merged**